### PR TITLE
add CAO cases

### PR DESCRIPTION
**Pull request by sunzhaohua2**
Automate 3 CAO cases. @jhou1 @miyadav please help to review, thanks.
https://openshift-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/Runner-v3/72746/console

### CODE REVIEW

File: lib/openshift/machine_set.rb [hunk 1]
```diff
@@ -15,6 +15,11 @@ def available_replicas(user: nil, cached: true, quiet: false)
       rr.dig('status', 'availableReplicas').to_i
     end
 
+    def machineset_name(user: nil, cached: true, quiet: false)
```
Comment (jhou1) [inline lib/openshift/machine_set.rb]: I think `machineset.name` can give you the same result.

**Comment by jhou1** [comment]
@sunzhaohua2 Thanks! The CAO tests are destructive, we need to discuss manage to restore the cluster after scenario.

I'm also giving a thought about automating CAO tests. What do you think if we clone a machineset and run the cluster-autoscaler tests against this cloned one?
Reason to suggest so it gives us an easier clean up -- by just deleting the autoscaler and the cloned machines/machinesets, we bring the cluster to the state before the test is run.

Something on my mind is like
```
Given I clone a machineset named "xxx"
And I run the :create admin command xxx (for cluster autoscaler)
And I run the :create admin command xxx (for machine autoscaler)

When I deploy workloads
Then the cluster should be scaled. 
```

**Comment by jhou1** [comment]
@sunzhaohua2 Some new steps are defined in https://github.com/openshift/verification-tests/pull/742 that you can use as a reference. The idea is that
- cluster autoscaler should be deployed for a cloned machineset
- With every cluster autoscaler or machine autoscaler created, it must be deleted after scenario
- Verify node number is the same after scenario
This is to ensure a consistent environment and do not cause failure to other tests

**Comment by sunzhaohua2** [comment]
@jhou1 Thanks. I will update.

**Comment by sunzhaohua2** [comment]
@jhou1 Have updated, please help to take a look. Local test result:
      [05:53:28] INFO> === End After Scenario: Update machineAutoscaler to reference a different MachineSet ===

1 scenario (1 passed)
32 steps (32 passed)
0m50.977s
[05:53:28] INFO> === At Exit ===

      [06:22:57] INFO> === End After Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource ===

1 scenario (1 passed)
9 steps (9 passed)
0m36.930s

      [06:20:45] INFO> === End After Scenario: Cao listens and deploys cluster-autoscaler based on ClusterAutoscaler resource ===

1 scenario (1 passed)
6 steps (6 passed)
0m41.250s
[06:20:45] INFO> === At Exit ===

File: features/machine/autoscaler.feature [hunk 2]
```diff
@@ -52,3 +52,104 @@ Feature: Cluster Autoscaler Tests
...
+  # @case_id OCP-21517
+  @admin
+  @destructive
+  Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+    Given I use the "openshift-machine-api" project
+    And I pick a random machineset to scale
+
+    When I run oc create over "https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/cloud/machine-autoscaler.yml" replacing paths:
```
Comment (jhou1) [inline features/machine/autoscaler.feature]: You didn't create the clusterautoscaler before creating the machineautoscaler, is the scenario valid?

File: features/machine/autoscaler.feature [hunk 3]
```diff
@@ -52,3 +52,104 @@ Feature: Cluster Autoscaler Tests
...
+
+  # @author zhsun@redhat.com
+  # @case_id OCP-21517
+  @admin
+  @destructive
+  Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+    Given I use the "openshift-machine-api" project
+    And I pick a random machineset to scale
```
Comment (jhou1) [inline features/machine/autoscaler.feature]: Please use the following steps to ensure you operate to a cloned machineset, and the cluster's node number is consistent before&after the test

```
    Given I store the number of machines in the :num_to_restore clipboard
    And admin ensures node number is restored to "<%= cb.num_to_restore %>" after scenario

    Given I clone a machineset named "machineset-clone-25436"
    And admin ensures "machineset-clone-25436" machineset is deleted after scenario
```

Comment (jhou1) [reply features/machine/autoscaler.feature]: you can use a different name for the cloned machineset

Comment (jhou1) [reply features/machine/autoscaler.feature]: Ok, I validated my self the clusterautoscaler is not needed

File: features/machine/autoscaler.feature [hunk 4]
```diff
@@ -52,3 +52,104 @@ Feature: Cluster Autoscaler Tests
...
+      | resource          | machineautoscaler                                                                |
+      | resource_name     | maotest0                                                                         |
+      | p                 | {"spec":{"scaleTargetRef":{"name":"<%= cb.machinesets[2].name %>"}}}             |
+      | type              | merge                                                                            |
+    Then the step should succeed
+    When I run the :describe admin command with:
+      | resource          | machineautoscaler                                                                |
+      | name              | maotest0                                                                         |
+    Then the step should succeed
+    And the output should match "Name:\s+<%= cb.machinesets[2].name %>"
```
Comment (jhou1) [inline features/machine/autoscaler.feature]: For this test, you assumed there are 3 machinesets in the cluster. Sometimes in an OSP deployment, there is only 1 machineset with 3 machines. The index used here will cause trouble.

Comment (jhou1) [reply features/machine/autoscaler.feature]: How about clone 2 machinesets and patch mao for them?

File: features/machine/autoscaler.feature [hunk 5]
```diff
@@ -52,3 +52,134 @@ Feature: Cluster Autoscaler Tests
...
+  @admin
+  @destructive
+  Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+
+    Given I store the number of machines in the :num_to_restore clipboard
+    And admin ensures node number is restored to "<%= cb.num_to_restore %>" after scenario
+    And admin ensures "machineset-clone" machineset is deleted after scenario
+    And admin ensures "maotest" machineautoscaler is deleted after scenario
```
Comment (jhou1) [inline features/machine/autoscaler.feature]: Please put each `admin ensures delete` right after their corresponding creation. This is for their read consistency.

For example:

```
Given I clone a machineset named "machineset-clone"
And admin ensures "machineset-clone" machineset is deleted after scenario
```

File: features/machine/autoscaler.feature (hunk 4 above)
Comment (sunzhaohua2) [reply features/machine/autoscaler.feature]: I cloned 3 machinesets, if clone 2 machineset, we could delete one mao and test another scenario.

      [16:16:20] INFO> === End After Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource ===

1 scenario (1 passed)
19 steps (19 passed)
7m2.305s

      [21:10:14] INFO> === End After Scenario: Update machineAutoscaler to reference a different MachineSet ===
1 scenario (1 passed)
42 steps (42 passed)
15m31.919s

File: features/machine/autoscaler.feature [hunk 6]
```diff
@@ -52,3 +52,134 @@ Feature: Cluster Autoscaler Tests
...
+      | ["metadata"]["name"]                   | maotest                    |
+      | ["spec"]["minReplicas"]                | 1                          |
+      | ["spec"]["maxReplicas"]                | 3                          |
+      | ["spec"]["scaleTargetRef"]["name"]     | <%= machine_set.name %>    |
+    Then the step should succeed
+    When I run the :get admin command with:
+      | resource          | machineautoscaler                               |
+      | resource_name     | maotest                                         |
+    Then the step succeeded
+    Then the expression should be true> machine_set.node_group_min(cached: false) == "1"
```
Comment (jhou1) [inline features/machine/autoscaler.feature]: Sorry I didn't know this earlier: https://github.com/openshift/verification-tests/pull/746/files#r387514585, maybe we don't have to define this method, sorry for the turn arounds

```
Then the expression should be true> machine_set.annotation("machine.openshift.io/cluster-api-autoscaler-node-group-min-size", cached: false) == "1"
```

File: features/machine/autoscaler.feature [hunk 7]
```diff
@@ -52,3 +52,134 @@ Feature: Cluster Autoscaler Tests
...
+      | name              | <%= cb.machineset_clone0_name %>                                                 |
+    Then the step should succeed
+    And the output should match "Annotations:\s+<none>"
+    When I run the :describe admin command with:
+      | resource          | machineset                                                                       |
+      | name              | <%= cb.machineset_clone2_name %>                                                 |
+    Then the step should succeed
+    And the output should match "Annotations:\s+autoscaling.openshift.io/machineautoscaler: openshift-machine-api/maotest0"
+  
+    When I run the :patch admin command with:
```
Comment (jhou1) [inline features/machine/autoscaler.feature]: Do the following tests duplicate the previous? I think you only need 2 machinesets and 1 machineautoscaler to cover the scenario.

Comment (sunzhaohua2) [reply features/machine/autoscaler.feature]: This case tests 2 scenarios:  
1) update a machineautoscaler to reference a  MachineSet not owned by another MachineAutoscaler
2) update a machineautoscaler to reference a  MachineSet already owned by another MachineAutoscaler
So 2 machineautuscalers are necessary, if we clone 2 machinesets we can cover the 2 scenarios by:
create mao1->machineset1
update mao1->machineset2
check result
create mao2->machineset1
update mao->machineset1
check result

Comment (jhou1) [reply features/machine/autoscaler.feature]: I get it, thank you.

**Comment by sunzhaohua2** [comment]
Have updated.
```
INFO> === End After Scenario: CAO listens and annotations machineSets based on MachineAutoscaler resource ===
1 scenario (1 passed)
19 steps (19 passed)
5m42.924s

      [07:03:01] INFO> === End After Scenario: Update machineAutoscaler to reference a different MachineSet ===
1 scenario (1 passed)
39 steps (39 passed)
11m24.569s
```

**Comment by jhou1** [comment]
@liangxia @pruan-rht Can PTAL? Thank you.

File: features/machine/autoscaler.feature [hunk 8]
```diff
@@ -52,3 +52,130 @@ Feature: Cluster Autoscaler Tests
...
+  Scenario: Cao listens and deploys cluster-autoscaler based on ClusterAutoscaler resource
+    Given I have an IPI deployment
+    And I switch to cluster admin pseudo user
+
+    When I run the :create admin command with:
+      | f | https://raw.githubusercontent.com/openshift-qe/v3-testfiles/master/cloud/cluster-autoscaler.yml   |
+    Then the step should succeed
+    And admin ensures "default" clusterautoscaler is deleted after scenario
+    And 1 pods become ready with labels:
+      | cluster-autoscaler=default,k8s-app=cluster-autoscaler                                                 |
```
Comment (liangxia) [inline features/machine/autoscaler.feature]: Bad indention

Comment (jhou1) [reply features/machine/autoscaler.feature]: Did you mean the table?

Comment (liangxia) [reply features/machine/autoscaler.feature]: Yes, the align I mean.  And aligns below.

Comment (sunzhaohua2) [reply features/machine/autoscaler.feature]: @jhou1 @liangxia Thanks, I have updated.

**Pull request merged**